megamek/src/megamek/client/ui/swing/lobby/ChatLounge.java

```diff
@@ -0,0 +1,3628 @@
+            }
+        };
+        
+        private void sorterPopup(MouseEvent e) {
+            ScalingPopup popup = new ScalingPopup();
+            GameOptions opts = clientgui.getClient().getGame().getOptions();
+            for (MekTableSorter sorter: union(unitSorters, bvSorters)) {
+                // Offer only allowed sorters and only one sorting direction
+                if (sorter.isAllowed(opts) && sorter.getSortingDirection() != Sorting.ASCENDING) {
+                    JMenuItem item = new JMenuItem(sorter.getDisplayName());
+                    item.addActionListener(mekTableHeaderAListener);
+                    item.setActionCommand(sorter.getDisplayName());
+                    popup.add(item);
+                }
+            }
+            popup.show(e.getComponent(), e.getX(), e.getY());
+        }
+
+    };
+    
+    /**
+     * Sets the sorting used in the Mek Table depending on the column header 
+     * that was clicked.  
+     */ 
+    private void changeMekTableSorter(MouseEvent e) {
+        int col = mekTable.columnAtPoint(e.getPoint());
+        MekTableSorter previousSorter = activeSorter;
+        List<MekTableSorter> sorters;
+        
+        // find the right list of sorters (or do nothing, if the column is not sortable)
+        if (col == MekTableModel.COL_UNIT) {
+            sorters = unitSorters;
+        } else if (col == MekTableModel.COL_BV) {
+            sorters = bvSorters;
+        } else {
+            return;
+        }
+        
+        // Select the next allowed sorter and refresh the display if the sorter was changed
+        nextSorter(sorters);
+        if (activeSorter != previousSorter) {
+            refreshMekTable();
+            updateTableHeaders();
+        }
+    }
+    
+    /** Selects the next allowed sorter in the given list of sorters. */
+    private void nextSorter(List<MekTableSorter> sorters) {
+        // Set the next sorter as active, if this column was already sorted, or
+        // the first sorter otherwise
+        int index = sorters.indexOf(activeSorter);
+        if (index == -1) {
+            activeSorter = sorters.get(0);
+        } else {
+            index = (index + 1) % sorters.size();
+            activeSorter = sorters.get(index);
+        }
+        
+        // Find an allowed sorter (e.g. blind drop may prohibit some)
+        int counter = 0; // Endless loop safeguard
+        while (!activeSorter.isAllowed(clientgui.getClient().getGame().getOptions())
+                && ++counter < 100) {
+            index = (index + 1) % sorters.size();
+            activeSorter = sorters.get(index);
+        }
+    }
+
+    /** Returns true when the compact view is active. */ 
+    public boolean isCompact() {
+        return butCompact.isSelected();
+    }
+    
+    /** 
+     * Returns a list of the selected entities in the Mek table. 
+     * The list may be empty but not null. 
+     */
+    private ArrayList<Entity> getSelectedEntities() {
```

Comment:
OK done, though I wonder at this practice (I also found it at some Java Practices website). Why stop at List? If I change the method to actually use a Set (which I've done multiple times while doing the lobby stuff) I still have to change the return value and possibly the callers. Isnt it the fault of the caller to not use a Collection whenever possible? If they do, I can return an ArrayList or a HashSet, no problem. Shouldnt the actual practice be to always receive a Collection?

Reply:
The reason I code using List/Set instead of Collection is that both give you more information and functionality while still being generic.

If you see a set, you know the values within are unique and that contains is a fast method to use.
If you see a list, you know the values within are ordered, sortable, accessible and insertable by index, and searchable by element.

The individual implementations within those two are less important, and can thus be genericized out, but just having a collection (especially with list) removes a huge amount of functionality and information for the programmer.

So... I'd use collection only if you don't and cannot think of a future reason why to use any list-based functionality and don't care about the duplication state (which... would just be easier to use and return List still to future proof it).